### PR TITLE
🐛 bug(Airbyte CDK): clean invalid fields from configured catalog

### PR DESCRIPTION
**Pull request by aldogonzalez8**
## What
Now that we can use [the selected fields from the configured catalog](https://github.com/airbytehq/airbyte-internal-issues/issues/2100) for example in [facebook-marketing](https://github.com/airbytehq/airbyte/pull/41039), we have found that there are cases where the configured catalog still has old fields that are no longer valid, causing problems [like this](https://airbytehq.sentry.io/issues/5212904883/events/bdaafc73518a40c38b8f84b52d322450/?project=6527718). 

```
Status:  400
  Response:
    {
      "error": {
        "message": "(#100) age_targeting, gender_targeting, labels, location, estimated_ad_recall_rate_lower_bound, estimated_ad_recall_rate_upper_bound, estimated_ad_recallers_lower_bound, estimated_ad_recallers_upper_bound are not valid for fields param after V19.0",
        "type": "OAuthException",
        "code": 100,
        "fbtrace_id": "Abgnrv0bw-v2MIobNaeELcH"
      }
    }
```

So, we need to do some work ourselves to validate that properties coming from the configured catalog are still valid in the current source schema.

## How
Let the CDK filter our invalid fields comparing with the result of `get_json_schema` method in the stream. 

## Review guide

1. `airbyte-cdk/python/airbyte_cdk/sources/streams/core.py`: new filter to validate configured schema against stream schema.
2. `airbyte-cdk/python/unit_tests/sources/streams/test_stream_read.py`: unit testing for this change.

## User Impact
When activating the latest version of source-Facebook-marketing, we expect not to keep seeing these invalid field errors from Facebook API.

## Can this PR be safely reverted and rolled back?

- [x] YES 💚
- [ ] NO ❌

### CODE REVIEW

**Comment by aldogonzalez8** [comment]
Manual testing

[Connection used to test locally](https://cloud.airbyte.com/workspaces/6705a0ab-302f-4175-813f-3f539241750a/connections/8d7631c6-edd4-4fe6-86ac-7d7efa3261f3/job-history)

Errors when running read with current cdk:

```
{"type":"LOG","log":{"level":"INFO","message":"During the sync, the following streams did not sync successfully: ads_insights: AirbyteTracedException('(#100) Tried accessing nonexisting field (cost_per_conversion_lead) on node type (AdsInsights)')"}}
{"type":"LOG","log":{"level":"FATAL","message":"None\nTraceback (most recent call last):\n  File \"<string>\", line 1, in <module>\n  File \"/Users/aldogonzalez/dev/airbyte/airbyte-integrations/connectors/source-facebook-marketing/source_facebook_marketing/run.py\", line 19, in run\n    launch(source, sys.argv[1:])\n  File \"/Users/aldogonzalez/dev/airbyte/airbyte-cdk/python/airbyte_cdk/entrypoint.py\", line 235, in launch\n    for message in source_entrypoint.run(parsed_args):\n  File \"/Users/aldogonzalez/dev/airbyte/airbyte-cdk/python/airbyte_cdk/entrypoint.py\", line 122, in run\n    yield from map(AirbyteEntrypoint.airbyte_message_to_string, self.read(source_spec, config, config_catalog, state))\n  File \"/Users/aldogonzalez/dev/airbyte/airbyte-cdk/python/airbyte_cdk/entrypoint.py\", line 166, in read\n    for message in self.source.read(self.logger, config, catalog, state):\n  File \"/Users/aldogonzalez/dev/airbyte/airbyte-cdk/python/airbyte_cdk/sources/abstract_source.py\", line 191, in read\n    raise AirbyteTracedException(message=error_message, failure_type=FailureType.config_error)\nairbyte_cdk.utils.traced_exception.AirbyteTracedException: None"}}
{"type":"TRACE","trace":{"type":"ERROR","emitted_at":1720827392174.356,"error":{"message":"During the sync, the following streams did not sync successfully: ads_insights: AirbyteTracedException('(#100) Tried accessing nonexisting field (cost_per_conversion_lead) on node type (AdsInsights)')","internal_message":null,"stack_trace":"Traceback (most recent call last):\n  File \"<string>\", line 1, in <module>\n  File \"/Users/aldogonzalez/dev/airbyte/airbyte-integrations/connectors/source-facebook-marketing/source_facebook_marketing/run.py\", line 19, in run\n    launch(source, sys.argv[1:])\n  File \"/Users/aldogonzalez/dev/airbyte/airbyte-cdk/python/airbyte_cdk/entrypoint.py\", line 235, in launch\n    for message in source_entrypoint.run(parsed_args):\n  File \"/Users/aldogonzalez/dev/airbyte/airbyte-cdk/python/airbyte_cdk/entrypoint.py\", line 122, in run\n    yield from map(AirbyteEntrypoint.airbyte_message_to_string, self.read(source_spec, config, config_catalog, state))\n  File \"/Users/aldogonzalez/dev/airbyte/airbyte-cdk/python/airbyte_cdk/entrypoint.py\", line 166, in read\n    for message in self.source.read(self.logger, config, catalog, state):\n  File \"/Users/aldogonzalez/dev/airbyte/airbyte-cdk/python/airbyte_cdk/sources/abstract_source.py\", line 191, in read\n    raise AirbyteTracedException(message=error_message, failure_type=FailureType.config_error)\nairbyte_cdk.utils.traced_exception.AirbyteTracedException: None\n","failure_type":"config_error","stream_descriptor":null}}}

```



After run with CDK changes in this PR

```
{"type":"LOG","log":{"level":"INFO","message":"Read 19 records from ads_insights stream"}}
{"type":"LOG","log":{"level":"INFO","message":"Marking stream ads_insights as STOPPED"}}
```

**Comment by aldogonzalez8** [comment]
CAT Failing for Zendesk in this PR is not related to CDK changes but to changes in the response of our account in Zendesk. [Please check this PR](https://github.com/airbytehq/airbyte/pull/41678).

File: airbyte-cdk/python/unit_tests/sources/declarative/partition_routers/test_substream_partition_router.py [hunk 1]
```diff
@@ -102,6 +102,9 @@ def read_records(
         if stream_slice and cursor_field and result:
             self._state["states"].append({cursor_field: result[-1][cursor_field], "partition": stream_slice["slice"]})
 
+    def get_json_schema(self) -> Mapping[str, Any]:
```
Comment (girarda) [inline airbyte-cdk/python/unit_tests/sources/declarative/partition_routers/test_substream_partition_router.py:105]: why is this method needed?

File: airbyte-cdk/python/airbyte_cdk/sources/streams/core.py [hunk 2]
```diff
@@ -544,4 +544,18 @@ def configured_json_schema(self) -> Optional[Dict[str, Any]]:
 
     @configured_json_schema.setter
     def configured_json_schema(self, json_schema: Dict[str, Any]) -> None:
-        self._configured_json_schema = json_schema
+        self._configured_json_schema = self._filter_schema_invalid_properties(json_schema)
+
+    def _filter_schema_invalid_properties(self, json_schema: Dict[str, Any | Dict[str, Any]]) -> Dict[str, Any]:
```
Comment (girarda) [inline airbyte-cdk/python/airbyte_cdk/sources/streams/core.py:549]: let's add a warning when filtering out fields so users know why they're not present in the destination

**Comment by aldogonzalez8** [comment]
**Manual testing** with connection id: 81d16659-48bf-44f3-ad98-17f4561e225e
- ads_insights
- ads_insights_action_type

In the master branch with the latest CDK changes (error expected):
`poetry run source-facebook-marketing read --config secrets/config.json --catalog integration_tests/configured_catalog.json`
```
{
  "type": "TRACE",
  "trace": {
    "type": "ERROR",
    "emitted_at": 1720986503890.427,
    "error": {
      "message": "Error code 100: (#100) age_targeting, gender_targeting, labels, location, estimated_ad_recall_rate_lower_bound, estimated_ad_recall_rate_upper_bound, estimated_ad_recallers_lower_bound, estimated_ad_recallers_upper_bound are not valid for fields param after V19.0.",
      "internal_message": "(#100) age_targeting, gender_targeting, labels, location, estimated_ad_recall_rate_lower_bound, estimated_ad_recall_rate_upper_bound, estimated_ad_recallers_lower_bound, estimated_ad_recallers_upper_bound are not valid for fields param after V19.0",
      "stack_trace": "Traceback (most recent call last):\n  ....",
      "failure_type": "system_error",
      "stream_descriptor": {
        "name": "ads_insights_action_type"
      }
    }
  }
}
{
  "type": "LOG",
  "log": {
    "level": "INFO",
    "message": "During the sync, the following streams did not sync successfully: ads_insights: AirbyteTracedException('(#100) age_targeting, gender_targeting, labels, location, estimated_ad_recall_rate_lower_bound, estimated_ad_recall_rate_upper_bound, estimated_ad_recallers_lower_bound, estimated_ad_recallers_upper_bound are not valid for fields param after V19.0')\nads_insights_action_type: AirbyteTracedException('(#100) age_targeting, gender_targeting, labels, location, estimated_ad_recall_rate_lower_bound, estimated_ad_recall_rate_upper_bound, estimated_ad_recallers_lower_bound, estimated_ad_recallers_upper_bound are not valid for fields param after V19.0')"
  }
}
```

Testing in fix branch `aldogonzalez8/cdk_to_remove_invalid_fields_from_configured_catalog` pointing to the cdk changes locally:
**pyproject.toml**
airbyte-cdk = {path = "../../../airbyte-cdk/python/", develop = true}

```bash
poetry lock 
poetry install
```

`poetry run source-facebook-marketing read --config secrets/config.json --catalog integration_tests/configured_catalog.json`

```
{
  "type": "LOG",
  "log": {
    "level": "INFO",
    "message": "Read 38 records from ads_insights stream"
  }
}
{
  "type": "LOG",
  "log": {
    "level": "INFO",
    "message": "Read 38 records from ads_insights_action_type stream"
  }
}
```

File: airbyte-cdk/python/unit_tests/sources/declarative/partition_routers/test_substream_partition_router.py (hunk 1 above)
Comment (aldogonzalez8) [reply airbyte-cdk/python/unit_tests/sources/declarative/partition_routers/test_substream_partition_router.py:105]: I found a few errors in the tests like the mock stream method was not implemented or unable to find the stream schema json file, so I just added a mock of the method as the purpose of these other tests was not to check something in the schema.

File: airbyte-cdk/python/airbyte_cdk/sources/streams/core.py (hunk 2 above)
Comment (aldogonzalez8) [reply airbyte-cdk/python/airbyte_cdk/sources/streams/core.py:549]: Sure, I'm adding the following, let me know if you have desire to rephrase:

[{
  "type": "LOG",
  "log": {
    "level": "WARN",
    "message": "For ads_insights_action_type the following properties in the configured catalog were not found in the the current stream schema and will be filtered: ['labels', 'location', 'age_targeting', 'gender_targeting', 'estimated_ad_recall_rate', 'estimated_ad_recallers_lower_bound', 'estimated_ad_recallers_upper_bound', 'estimated_ad_recall_rate_lower_bound', 'estimated_ad_recall_rate_upper_bound']"
  }
}](url)

File: airbyte-cdk/python/unit_tests/sources/declarative/partition_routers/test_substream_partition_router.py (hunk 1 above)
Comment (aldogonzalez8) [reply airbyte-cdk/python/unit_tests/sources/declarative/partition_routers/test_substream_partition_router.py:105]: For example, here, I got this error in several tests before mock `get_json_schema`:

```
self = <[AttributeError("'MockStream' object has no attribute 'retriever'") raised in repr()] MockStream object at 0x108afc3d0>

    def get_json_schema(self) -> Mapping[str, Any]:  # type: ignore
        """
        :return: A dict of the JSON schema representing this stream.
    
        The default implementation of this method looks for a JSONSchema file with the same name as this stream's "name" property.
        Override as needed.
        """
>       return self._schema_loader.get_json_schema()
E       AttributeError: 'MockStream' object has no attribute '_schema_loader'. Did you mean: 'schema_loader'?
```

File: airbyte-cdk/python/airbyte_cdk/sources/streams/core.py [hunk 3]
```diff
@@ -544,4 +544,28 @@ def configured_json_schema(self) -> Optional[Dict[str, Any]]:
```
Comment (girarda) [inline airbyte-cdk/python/airbyte_cdk/sources/streams/core.py:544]: @aldogonzalez8 we should implement this filtering at the CDK level so other connectors also benefit from the improvement. Is there a way to move this method's logic to `core.py`?

**Review by girarda — COMMENTED** [review]
@aldogonzalez8 the change looks good, but we should implement this filter in the CDK so other connectors also benefit from it.

File: airbyte-cdk/python/airbyte_cdk/sources/streams/core.py (hunk 3 above)
Comment (girarda) [reply airbyte-cdk/python/airbyte_cdk/sources/streams/core.py:544]: nope. that's my bad. this is absolutely right <3

File: airbyte-cdk/python/airbyte_cdk/sources/streams/core.py [hunk 4]
```diff
@@ -544,4 +544,28 @@ def configured_json_schema(self) -> Optional[Dict[str, Any]]:
...
+        Configured Schemas can have very old fields, so we need to housekeeping ourselves.
+        """
+        received_schema_properties: Any = json_schema.get("properties", {})
+        stream_schema_properties: Any = self.get_json_schema().get("properties", {})
+        valid_configured_schema_properties = {
+            stream_property: property_value
+            for stream_property, property_value in received_schema_properties.items()
+            if stream_property in stream_schema_properties
+        }
+        invalid_properties = []
```
[the remark this answers is not in the document]
Comment (aldogonzalez8) [reply airbyte-cdk/python/airbyte_cdk/sources/streams/core.py:561]: Yes, but you would still need to go through stream_schema_properties to get nested data.
```
            "cpc": {
              "type": ["null", "number"]
            },
```

[the remark this answers is not in the document]
Comment (aldogonzalez8) [reply airbyte-cdk/python/airbyte_cdk/sources/streams/core.py:561]: Doing small renaming of `valid_configured_schema_properties` to `valid_configured_schema_properties_keys`:
```
configured_keys = received_schema_properties.keys() 
stream_keys = stream_schema_properties.keys() 

invalid_properties = configured_keys - stream_keys 
valid_configured_schema_properties_keys = stream_keys & configured_keys
```

Example of neste data we want to keep (types):
```
            "cpc": {
              "type": ["null", "number"]
            },
```

```python
for configured_schema_property in valid_configured_schema_properties_keys:
   valid_configured_schema_properties[configured_schema_property] = stream_schema_properties[configured_schema_property]
```

[the remark this answers is not in the document]
Comment (aldogonzalez8) [reply airbyte-cdk/python/airbyte_cdk/sources/streams/core.py:561]: I will add a few extra lines to the unit test to ensure that besides the invalid properties, no other keys are also removed and the inner data remains the same.

[the remark this answers is not in the document]
Comment (aldogonzalez8) [reply airbyte-cdk/python/airbyte_cdk/sources/streams/core.py:561]: Manual testing locally after applying these changes still looks good on one of the connections that was failing on Friday:
```json
{
  "type": "STATE",
  "state": {
    "type": "STREAM",
    "stream": {
      "stream_descriptor": {
        "name": "ads_insights",
        "namespace": null
      },
      "stream_state": {
        "602863491279135": {
          "date_start": "2024-07-15",
          "slices": []
        },
        "time_increment": 1
      }
    },
    "sourceStats": {
      "recordCount": 19.0
    }
  }
}

{
  "type": "STATE",
  "state": {
    "type": "STREAM",
    "stream": {
      "stream_descriptor": {
        "name": "ads_insights_action_type",
        "namespace": null
      },
      "stream_state": {
        "602863491279135": {
          "date_start": "2024-07-15",
          "slices": []
        },
        "time_increment": 1
      }
    },
    "sourceStats": {
      "recordCount": 19.0
    }
  }
}

```

[the remark this answers is not in the document]
Comment (aldogonzalez8) [reply airbyte-cdk/python/airbyte_cdk/sources/streams/core.py]: Actually, if there are no invalid fields, let's just return the very same configured catalog and skip filtering; it doesn't make sense to continue.

**Review by girarda — APPROVED** [review]
looks good thanks for the quick turnaround @aldogonzalez8 !

**Pull request merged**